powertools-sqs/src/main/java/software/amazon/lambda/powertools/sqs/SqsUtils.java

```diff
@@ -489,15 +493,41 @@ public static <R> List<R> batchProcessor(final SQSEvent event,
             client = SqsClient.create();
         }
 
+        // If we are working on a FIFO queue, when any message fails we should stop processing and return the
+        // rest of the batch as failed too. We use this variable to track when that has happened.
+        // https://docs.aws.amazon.com/lambda/latest/dg/with-sqs.html#services-sqs-batchfailurereporting
+        boolean failWholeBatch = false;
+
         BatchContext batchContext = new BatchContext(client);
 
         for (SQSMessage message : event.getRecords()) {
-            try {
-                handlerReturn.add(handler.process(message));
-                batchContext.addSuccess(message);
-            } catch (Exception e) {
-                batchContext.addFailure(message, e);
-                LOG.error("Encountered issue processing message: {}", message.getMessageId(), e);
+            // If the batch has already failed, we mark each message past the failure point as failed
+            if (failWholeBatch) {
```

Comment:
Would it be possible to avoid looping over the remaining messages? Once it is failed, we add the other messages in the failures in one step after the loop (I did it in the [RFC](https://github.com/aws-powertools/powertools-lambda-java/blob/rfc-batch-processor/powertools-batch/src/main/java/software/amazon/lambda/powertools/batch2/SQSBatchProcessor.java#L75-L103))

Reply:
I'm not so sure about this - I find that variant quite hard to grok with the tracking of the offset into the list and the c-style cursor movement. I think we should err on the side of clarity as we're dealing with small lists. I could imagine another alternative could be:

1. Push everything into a queue
2. Pop one by one until you fail
3. Once you fail `break;` then push the rest into the failure list

WDYT @jeromevdl ?

Reply:
I kinda like what I did as an optimization... do whatever you prefer but please don't loop over failed message. (I find strange to push message into a queue, while they are already messages from a queue ;) )

Reply:
This should be the last detail @jeromevdl. I've tried to make the behaviour explicit, if verbose. I agree that shifting a pointer around will be more efficient. The maximum batch size of a regular queue is 10k or 10 for a FIFO queue. Maybe we should actually do it your way.

Reply:
I read more and agree with you. 10k messages per batch maximum I'm happy to shift focus to efficiency. 

Here's the loop:

https://github.com/aws-powertools/powertools-lambda-java/blob/685177605bc9d5a96f4a1ea96102679c06e9475b/powertools-sqs/src/main/java/software/amazon/lambda/powertools/sqs/SqsUtils.java#L498-L529

And here's capturing the failures:

https://github.com/aws-powertools/powertools-lambda-java/blob/685177605bc9d5a96f4a1ea96102679c06e9475b/powertools-sqs/src/main/java/software/amazon/lambda/powertools/sqs/SqsUtils.java#L533-L541

@jeromevdl I think this should be it! Please have a bit of a dive in here, but I think the test coverage is good (and I added an extra one for this).